ginkgo/internal/suite.go

```diff
@@ -0,0 +1,176 @@
+		closed:            false,
+	}
+}
+
+func (s *SuiteTest) Context() context.Context {
+	return s.state.ctx
+}
+
+func (s *SuiteTest) EnterContainer(stepName string, text string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.closed {
+		return errors.New("suite is closed")
+	}
+
+	span, ctx := tracer.StartSpanFromContext(s.state.ctx, text,
+		tracer.Tag("test.name", text),
+		tracer.Tag("test.type", "test"),
+		tracer.Tag("test.framework", s.TestFrameworkName),
+		tracer.Tag("ginkgo.seed", ginkgo.GinkgoRandomSeed()),
+		tracer.Tag("ginkgo.step", stepName),
+	)
+
+	s.state.Push(WithTestCount(WithFinish(ctx, func() { span.Finish() }), 0))
+
+	return nil
+}
+
+func (s *SuiteTest) LeaveContainer() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.closed {
+		return errors.New("suite is closed")
+	}
+
+	if err := s.state.Pop(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type test struct {
+	ctx     context.Context
+	finish  func()
+	options []ddtesting.Option
+}
+
+func newTest(ctx context.Context, opts ...ddtesting.Option) *test {
+	for ctx := ctx; ctx != nil; ctx = GetParent(ctx) {
+		if IncTestCount(ctx) < 1 {
+			panic("testContextProgress.push: corrupted stack")
+		}
+	}
+
+	debugContextSize(ctx)
+
+	return &test{ctx: ctx, options: opts}
+}
+
+func (t *test) Enter(stepName, suiteName string) error {
+	opts := append(t.options, ddtesting.WithSpanOptions(
+		tracer.Tag("test.type", "test"),
+		tracer.Tag("test.suite", suiteName),
+		tracer.Tag("ginkgo.step", stepName),
+		tracer.Tag("ginkgo.seed", ginkgo.GinkgoRandomSeed()),
+		tracer.Tag("ginkgo.parallelProcess", ginkgo.GinkgoParallelProcess()),
+		tracer.Tag("ginkgo.numAttempts", reporting.CurrentSpecReport().NumAttempts),
+	))
+
+	_, t.finish = ddtesting.StartTestWithContext(t.ctx, ginkgo.GinkgoT(1), opts...)
+
+	return nil
+}
+
+func (t *test) Leave() error {
+	if t.finish == nil {
+		return errors.New("test is not entered")
+	}
+
+	t.finish()
+
+	nbToPop := 0
+
+	defer debugContextSize(t.ctx)
+
+	for ctx := t.ctx; ctx != nil; ctx = GetParent(ctx) {
+		value := DecTestCount(ctx)
+
+		switch {
+		case value == 0:
+			if nbToPop < 0 {
+				return errors.New("corrupted stack")
+			}
+
+			nbToPop++
+
+			if err := Finish(ctx); err != nil {
+				panic(err)
+			}
+		case value < 0:
+			return fmt.Errorf("count < 0 for %+v", ctx)
+		default:
+			nbToPop = -1 // mark the stack as corrupted
+		}
+	}
+
+	return nil
+}
+
+type TestCase interface {
+	Enter(stepName, suiteName string) error
+	Leave() error
+}
+
+type Snapshot struct {
+	FrameworkName string
+	ctx           context.Context
+}
+
+func (s *SuiteTest) Snapshot() *Snapshot {
+	return &Snapshot{
+		FrameworkName: s.TestFrameworkName,
+		ctx:           s.state.ctx,
+	}
+}
+
+func (s *Snapshot) RegisterTest() TestCase {
+	return newTest(
+		s.ctx,
+		ddtesting.WithSpanOptions(tracer.Tag("test.framework", s.FrameworkName)),
+	)
+}
+
+func (s *SuiteTest) Close() error {
```

Comment:
Future improvement: we don't yet have suite-level visibility in dd-sdk-go-testing but this would be the place to send a `test_suite_end` event with the duration of the whole suite.